be/src/storage/lake/local_pk_index_manager.cpp

```diff
@@ -0,0 +1,161 @@
+            } else if (!tablet_manager->is_tablet_in_worker(id)) {
+                // the shard may be scheduled to other nodes
+                if (!update_manager->try_lock_pk_index_shard(id)) {
+                    LOG(WARNING) << "Fail to lock pk index, tablet id: " << id;
+                    continue;
+                }
+                not_in_worker_tablet_ids.emplace_back(id);
+                // judge whether tablet is scheduled again,
+                // and pk_index_shard write_lock has been hold, so no process will build the persistent index.
+                if (!tablet_manager->is_tablet_in_worker(id)) {
+                    // try to remove pk index cache to avoid continuing to use the index in the cache after deletion.
+                    if (update_manager->try_remove_primary_index_cache(id)) {
+                        if (StorageEngine::instance()->clear_persistent_index(data_dir, id, tablet_pk_path).ok()) {
+                            removed_dir_tablet_ids.push_back(id);
+                        }
+                    }
+                }
+                update_manager->unlock_pk_index_shard(id);
+            }
+        }
+        int64_t t_end = MonotonicMillis();
+        LOG(INFO) << "finish gc local persistent index dir: " << pk_path
+                  << ", found tablet not in the worker, tablet_ids: " << JoinInts(not_in_worker_tablet_ids, ",")
+                  << ", data_dir changed tablet_ids: " << JoinInts(dir_changed_tablet_ids, ",")
+                  << ", and removed dir successfully, tablet_ids: " << JoinInts(removed_dir_tablet_ids, ",")
+                  << ", cost:" << t_end - t_start << "ms";
+    }
+}
+
+void LocalPkIndexManager::evict(UpdateManager* update_manager,
+                                std::unordered_map<DataDir*, std::set<std::string>>& store_to_tablet_ids) {
+    for (const auto& [data_dir, tablet_ids] : store_to_tablet_ids) {
+        bool need_evict = false;
+        auto space_info_or = FileSystem::Default()->space(data_dir->path());
+        if (space_info_or.ok()) {
+            auto space_info = *space_info_or;
+            need_evict = (double)space_info.free < (double)space_info.capacity * config::starlet_cache_evict_low_water;
+        }
+        TEST_SYNC_POINT_CALLBACK("LocalPkIndexManager::evict:1", &need_evict);
+        if (!need_evict) {
+            continue;
+        }
+
+        auto pk_path = data_dir->get_persistent_index_path();
+        LOG(INFO) << "start to evict local persistent index dir:" << pk_path;
+        int64_t t_start = MonotonicMillis();
+
+        std::vector<int64_t> tablet_ids_to_be_evicted;
+        std::vector<int64_t> removed_dir_tablet_ids;
+
+        for (const auto& tablet_id : tablet_ids) {
+            auto tablet_pk_path = pk_path + "/" + tablet_id;
+            int64_t id = std::stoll(tablet_id);
+            std::set<std::string> index_files;
+            Status ret = fs::list_dirs_files(tablet_pk_path, nullptr, &index_files);
```

Comment:
List directory only after you have finished checking the modification time of the directory.